File: fruits.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        
        unordered_map<int, int> map;
        
        int size = 0, max = 0;
        
        for(int windowStart = 0, windowEnd = 0, n = fruits.size(); windowEnd < n; windowEnd++)
        {
            map[fruits[windowEnd]]++;
            
            while(map.size() > 2)
            {
                map[fruits[windowStart]]--;
                
                if(map[fruits[windowStart]] == 0)  map.erase(fruits[windowStart]);
                
                windowStart++;
            }
            
            size = windowEnd - windowStart + 1;
            
            if(size > max) max = size;
        }
        
        return max;
    }
};

File: progress_sheet/fruits.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        
        left = 0
        size = len(fruits)
        collected = defaultdict(int)
        max_fruits = 0

        # find the maximum fruits that can be collected
        for right in range(size):
                        
            collected[fruits[right]] += 1

            # we can only collect two types of fruits
            while len(collected) > 2:
                collected[fruits[left]] -= 1

                if collected[fruits[left]] == 0:
                    del collected[fruits[left]]

                left += 1
            
            max_fruits = max(max_fruits, right - left + 1)

        return max_fruits
